test/sqllogictest/column_knowledge.slt

```diff
@@ -0,0 +1,688 @@
+| Filter (#0 = 123)
+
+%1 =
+| Get %0 (l0)
+| ArrangeBy ()
+
+%2 =
+| Get materialize.public.t2 (u3)
+| Filter (#0 = 123)
+
+%3 = Let l1 =
+| Join %1 %2
+| | implementation = Differential %2 %1.()
+| | demand = (#0, #1, #3)
+
+%4 =
+| Get %3 (l1)
+| Distinct group=(123, #1)
+| Negate
+
+%5 =
+| Get %0 (l0)
+| Distinct group=(123, #1)
+
+%6 =
+| Union %4 %5
+| ArrangeBy (#1)
+
+%7 =
+| Join %6 %0 (= #1 #3)
+| | implementation = Differential %0 %6.(#1)
+| | demand = (#0, #1)
+| Map null, null
+| Project (#0, #1, #4, #5)
+
+%8 =
+| Union %3 %7
+| Project (#0, #1, #3)
+
+EOF
+
+query T multiline
+EXPLAIN SELECT * FROM t1 LEFT JOIN t2 ON (TRUE) WHERE t1.f1 = t2.f1 AND t1.f1 = 123;
+----
+%0 =
+| Get materialize.public.t1 (u1)
+| Filter (#0 = 123)
+| ArrangeBy ()
+
+%1 =
+| Get materialize.public.t2 (u3)
+| Filter (#0 = 123)
+
+%2 =
+| Join %0 %1
+| | implementation = Differential %1 %0.()
+| | demand = (#0..#3)
+
+EOF
+
+# Transitive application
+
+query T multiline
+EXPLAIN SELECT * FROM t1, t2, t3 WHERE t1.f1 = 123 AND t1.f1 = t2.f1 AND t2.f1 = t3.f1;
+----
+%0 =
+| Get materialize.public.t1 (u1)
+| Filter (#0 = 123)
+| ArrangeBy ()
+
+%1 =
+| Get materialize.public.t2 (u3)
+| Filter (#0 = 123)
+| ArrangeBy ()
+
+%2 =
+| Get materialize.public.t3 (u5)
+| Filter (#0 = 123)
+
+%3 =
+| Join %0 %1 %2
+| | implementation = Differential %2 %0.() %1.()
+| | demand = (#0..#5)
+
+EOF
+
+# HAVING clause
+
+query T multiline
+EXPLAIN SELECT t1.f1 FROM t1, t2 WHERE t1.f1 = t2.f1 GROUP BY t1.f1 HAVING t1.f1 = 123;
+----
+%0 =
+| Get materialize.public.t1 (u1)
+| Filter (#0 = 123)
+| ArrangeBy ()
+
+%1 =
+| Get materialize.public.t2 (u3)
+| Filter (#0 = 123)
+
+%2 =
+| Join %0 %1
+| | implementation = Differential %1 %0.()
+| | demand = (#0)
+| Project (#0)
+
+EOF
+
+#
+# Subqueries
+#
+
+query T multiline
+EXPLAIN SELECT (SELECT t1.f1 FROM t1 WHERE t1.f1 = t2.f1) FROM t2 WHERE t2.f1 = 123;
+----
+%0 = Let l0 =
+| Get materialize.public.t2 (u3)
+| Filter (#0 = 123)
+
+%1 = Let l1 =
+| Get %0 (l0)
+| Distinct group=(123)
+
+%2 =
+| Get %1 (l1)
+| ArrangeBy ()
+
+%3 =
+| Get materialize.public.t1 (u1)
+| Filter (#0 = 123)
+
+%4 = Let l2 =
+| Join %2 %3
+| | implementation = Differential %3 %2.()
+| | demand = (#1)
+
+%5 =
+| Get %4 (l2)
+| Project (#0, #1)
+
+%6 =
+| Get %4 (l2)
+| Reduce group=(123)
+| | agg count(true)
+| Filter (#1 > 1)
+| Map (err: more than one record produced in subquery)
+| Project (#0, #2)
+
+%7 = Let l3 =
+| Union %5 %6
+
+%8 =
+| Get %0 (l0)
+| ArrangeBy ()
+
+%9 =
+| Get %7 (l3)
+| Distinct group=(123)
+| Negate
+
+%10 =
+| Union %9 %1
+| ArrangeBy ()
+
+%11 =
+| Join %10 %1
+| | implementation = Differential %1 %10.()
+| | demand = ()
+| Map null
+| Project (#0, #2)
+
+%12 =
+| Union %7 %11
+
+%13 =
+| Join %8 %12
+| | implementation = Differential %12 %8.()
+| | demand = (#3)
+| Project (#3)
+
+EOF
+
+# This case is currently not optimized
+query T multiline
+EXPLAIN SELECT (SELECT t1.f1 FROM t1) = t2.f1 FROM t2 WHERE t2.f1 = 123;
```

Comment:
This is one of the few cases that I found where propagation does not kick in. I think this is OK, given the sheer number of other cases that are now optimized.

Reply:
I'm not sure what you mean by "propagation does not kick in". Didn't propagation kick in because the Map in %8 reads `(#2 = 123)` instead of `(#2 = #0)`?

Reply:
I thought the ```Filter (#0 = 123)``` should arrive all the way to %0 and %1